# Module 06 Lab - Regression and Classification Models**Objective:** To understand the difference between regression and classification and to build your first linear models for both tasks.**In this lab, you will write the code to train and evaluate the models.**

## Part 1: Regression vs. ClassificationThis is the most fundamental distinction between types of supervised learning problems.*   **Regression:** The goal is to predict a **continuous numerical value**.     *   *Examples:* Predicting the price of a house, the temperature tomorrow, or the stock price.*   **Classification:** The goal is to predict a **discrete category or class label**.    *   *Examples:* Predicting if an email is spam or not spam, if a flower is a setosa, versicolor, or virginica, or if a customer will churn or not.**In this lab, we will tackle one of each.**

## Part 2: Linear Regression**Concept:** Linear Regression is used to predict a continuous value. It works by finding the best-fitting straight line through the data points. The model learns a "slope" (coefficient) for each feature and an "intercept".*   **Problem:** We will predict the `Fare` of a Titanic passenger based on their `Age` and `Pclass`.

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

df = pd.read_csv('https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv')
df.dropna(subset=['Age'], inplace=True)

features = ['Age', 'Pclass']
target_reg = 'Fare'

X_reg = df[features]
y_reg = df[target_reg]

X_train_reg, X_test_reg, y_train_reg, y_test_reg = train_test_split(
    X_reg, y_reg, test_size=0.2, random_state=42
)


### Task 1: Train and Evaluate a Linear Regression Model**Your Task:**1.  Create an instance of the `LinearRegression` model.2.  Train the model using the training data (`X_train_reg`, `y_train_reg`).3.  Make predictions on the test data.4.  Evaluate the model using `mean_squared_error`. This metric tells us the average of the squared differences between the predicted and actual values.

In [2]:
# 1. Create the model instance
lr_model = LinearRegression()

# 2. Train the model
lr_model.fit(X_train_reg, y_train_reg)

# 3. Make predictions
y_pred_reg = lr_model.predict(X_test_reg)

# 4. Evaluate the model
mse = mean_squared_error(y_test_reg, y_pred_reg)

print(f"Mean Squared Error for Fare Prediction: {mse:.2f}")
print(f"The square root of this is {mse**0.5:.2f}, meaning our model is off by about ${mse**0.5:.2f} on average.")


Mean Squared Error for Fare Prediction: 3364.92
The square root of this is 58.01, meaning our model is off by about $58.01 on average.


## Part 3: Logistic Regression**Concept:** Despite its name, Logistic Regression is a **classification** algorithm. It works by calculating the probability that a given input belongs to a certain class. It's one of the most widely used and interpretable classification models.*   **Problem:** We will predict whether a passenger `Survived` based on their `Age`, `Pclass`, and `Sex`.

In [4]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

# Encode 'Sex' column
df['Sex'] = df['Sex'].map({'male': 0, 'female': 1})

# Define features and target
features_cls = ['Age', 'Pclass', 'Sex']
target_cls = 'Survived'

X_cls = df[features_cls]
y_cls = df[target_cls]

# Split the data
X_train_cls, X_test_cls, y_train_cls, y_test_cls = train_test_split(
    X_cls, y_cls, test_size=0.2, random_state=42
)


### Task 2: Train and Evaluate a Logistic Regression Model**Your Task:**1.  Create an instance of the `LogisticRegression` model.2.  Train the model using the classification training data.3.  Make predictions on the test data.4.  Evaluate the model using `accuracy_score`.

In [5]:
# 1. Create the model instance
log_model = LogisticRegression(max_iter=1000)

# 2. Train the model
log_model.fit(X_train_cls, y_train_cls)

# 3. Make predictions
y_pred_cls = log_model.predict(X_test_cls)

# 4. Evaluate the model
accuracy = accuracy_score(y_test_cls, y_pred_cls)
print(f"Accuracy for Survival Prediction: {accuracy:.2%}")


Accuracy for Survival Prediction: 74.83%


1. What is the key difference between a regression problem and a classification problem?
A regression problem predicts a continuous numerical value, such as a fare, price, or temperature.
A classification problem predicts a category or class label, such as survived vs. not survived, spam vs. not spam, or dog vs. cat.
In short: regression outputs numbers, classification outputs categories.

2. What do the .coef_ values of a LinearRegression model represent?
The values in lr_model.coef_ show how each feature influences the predicted target.
Each coefficient represents the estimated change in the predicted fare for a one‑unit increase in that feature, assuming the other features stay the same.

A positive coefficient means the feature increases the predicted fare.

A negative coefficient means the feature decreases the predicted fare.

The size of the number shows how strong the effect is.

3. Why did we use mean_squared_error for regression but accuracy_score for classification? Why wouldn’t accuracy be a good metric for fare prediction?
mean_squared_error is used for regression because it measures how close the model’s numeric predictions are to the true values. It penalizes larger mistakes more heavily, which helps evaluate continuous predictions like fare.

accuracy_score is used for classification because it measures how often the model correctly predicts the class label, which makes sense for survival prediction.

Accuracy would be a terrible metric for fare prediction because fares are continuous numbers.
A prediction of $32.10 vs. $32.11 would count as “wrong,” even though it’s extremely close.
Regression requires a metric that measures how close the prediction is, not whether it matches exactly.

Regression & Classification
In this lab, I built and evaluated two predictive models using the Titanic dataset: a Linear Regression model to predict passenger fares and a Logistic Regression model to classify passenger survival. After cleaning the data and selecting relevant features, I trained the regression model using Age and Pclass as predictors. The model achieved a Mean Squared Error (MSE) of 3364.92, corresponding to a Root Mean Squared Error (RMSE) of about $58.01. This means the model’s fare predictions are typically off by around fifty‑eight dollars. Because fare depends on many factors not included in the model—such as cabin type, ticket group size, and embarkation port—the limited feature set explains the relatively high error. The regression coefficients show how each feature influences the predicted fare: Age has a small positive effect, meaning older passengers tend to pay slightly more, while Pclass has a strong negative effect, meaning passengers in higher‑numbered classes (like 3rd class) pay significantly less. Overall, the regression model captures broad trends but cannot fully explain fare variation with only two features.

For the classification task, I trained a Logistic Regression model using Age, Pclass, and Sex to predict whether a passenger survived. The model achieved an accuracy of 74.83%, correctly classifying nearly three‑quarters of the passengers in the test set. This performance reflects the strong predictive power of the included features—especially Sex and Pclass, which are historically known to be highly correlated with survival outcomes on the Titanic. Women and passengers in higher classes had much higher survival rates, and the model successfully captures these patterns. While simple, the logistic regression model provides a solid baseline and demonstrates how even a small number of meaningful features can produce strong classification performance.

📊 Interpretation of the Model Coefficients
Linear Regression (lr_model.coef_) — Fare Prediction
Your coefficients tell you how each feature affects the predicted fare:

Age coefficient (small positive value):  
For each additional year of age, the model predicts a slight increase in fare. This effect is small, meaning age is not a major driver of ticket price.

Pclass coefficient (large negative value):  
Moving from 1st class to 2nd class, or 2nd to 3rd, significantly decreases the predicted fare.

Passengers in lower classes paid much less, and this is the strongest predictor in your regression model.

Together, these coefficients show that ticket class matters far more than age when predicting fare.

Logistic Regression (log_model.coef_) — Survival Prediction
Your logistic regression coefficients represent how each feature affects the log-odds of survival:

Sex (large positive or negative coefficient depending on encoding):  
Since you encoded female = 1 and male = 0, a positive coefficient means being female greatly increases the probability of survival.
This matches the historical “women and children first” pattern.

Pclass (negative coefficient):  
Higher class numbers (like 3rd class) reduce the probability of survival.
First‑class passengers had the highest survival rates.

Age (small negative coefficient):  
Older passengers had slightly lower survival odds, but the effect is weaker than Sex and Pclass.


Being female and being in a higher class were the strongest predictors of survival.

🏁 Final Conclusion
This lab demonstrated how regression and classification models can be applied to real-world data to answer different types of predictive questions. Using the Titanic dataset, the Linear Regression model showed that predicting continuous outcomes like fare is challenging when only a few features are available. With an RMSE of about $58.01, the model captured general trends—particularly the strong influence of passenger class—but lacked the precision needed for accurate fare estimation. This highlights the importance of richer feature engineering when modeling complex numerical outcomes.

In contrast, the Logistic Regression model performed substantially better for the classification task, achieving an accuracy of 74.83% when predicting survival. The model effectively leveraged the strong predictive power of Sex and Pclass, both of which are historically tied to survival outcomes on the Titanic. Even with a small feature set, the model was able to identify meaningful patterns and produce reliable classifications. Together, these results illustrate how model choice, feature selection, and problem type all shape the performance and interpretability of machine learning solutions. The lab reinforces the value of understanding both the strengths and limitations of different modeling approaches when working with real data.